### PR TITLE
[BUG][Data] Support for custom filters with heterogeneous data fields

### PR DESCRIPTION
**Pull request by kavilla**
### Description

When enabling the advanced setting `courier:ignoreFilterIfFieldNotInIndex` Custom OpenSearch Query DSL filters could technically be applied to index patterns that map to indices that are not exactly the same. Since the custom query filter is a user input then users can really type anything that they need. Or any field that they know is present but we do not know for sure.

Therefore, we can check if the id which is the index pattern title to check if we should apply the filter or not.

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5581
I believe issue:
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5423

Should closed as that is expected functionality.

## Screenshot

Request before:
Screenshot 2023-12-06 at 2 16 05 AM

Request after:
Screenshot 2023-12-06 at 2 19 56 AM

## Testing the changes

Using the sample data web logs and created a data table. 

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by ananzh — APPROVED** [review]

**Comment by ananzh** [comment]
Issue Resolve https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5423

File: src/plugins/data/common/opensearch_query/opensearch_query/filter_matches_index.ts [hunk 1]
```diff
@@ -31,14 +31,14 @@
 import { IIndexPattern, IFieldType } from '../../index_patterns';
 import { Filter } from '../filters';
 
-/*
- * TODO: We should base this on something better than `filter.meta.key`. We should probably modify
- * this to check if `filter.meta.index` matches `indexPattern.id` instead, but that's a breaking
- * change.
- */
```
Comment (joshuarrrr) [inline src/plugins/data/common/opensearch_query/opensearch_query/filter_matches_index.ts:38]: Let's make an issue that captures the uncertainty we still have here, where we can discuss the other underlying issues, including lack of validation on the custom filter inputs.

**Review by joshuarrrr — APPROVED** [review]

**Comment by kavilla** [comment]
> What do we think about backporting this to 1.3?

i think so

**Comment by kavilla** [comment]
Verified newest commit locally.

**Review by ZilongX — APPROVED** [review]

**Pull request merged**